src/privileged/vpnRecommender/api.js

```diff
@@ -0,0 +1,537 @@
+  }
+
+  sendTelemetry(data) {
+    if (this.telemetryCallback) {
+      this.telemetryCallback(data);
+    }
+  }
+
+  dontShowChecked(messageName) {
+    Preferences.set(DONT_SHOW_PREF, true);
+
+    this.telemetryCallback({
+      "message_type": "event",
+      "event": "dont-show-checked",
+      "message-name": messageName,
+    });
+  }
+
+  openVpnPage() {
+    const variation = this.variation;
+    const urlArgs = Object.assign({}, VPN_LANDING_PAGE_DEFAULT_PARAMS,
+      {"utm_content": variation});
+
+    const mergeQueryArgs = (url, ...args) => {
+      /* currently left to right*/
+      const U = new URL(url);
+      let q = U.search || "?";
+      q = new URLSearchParams(q);
+
+      const merged = Object.assign({}, ...args);
+
+      // get user info.
+      Object.keys(merged).forEach((k) => {
+        q.set(k, merged[k]);
+      });
+
+      U.search = q.toString();
+      return U.toString();
+    };
+
+    const win = this.RecentWindow.getMostRecentBrowserWindow();
+    const tab = win.gBrowser.addWebTab(mergeQueryArgs(VPN_LANDING_PAGE_URL, urlArgs));
+    win.gBrowser.selectedTab = tab;
+  }
+
+  notificationActionCallback(message) {
+    log(`notification action: name => ${message.name}, data=> ${JSON.stringify(message.data)}`);
+
+    let eventName;
+
+    if (message.data && message.data.dontShowChecked) {
+      this.dontShowChecked(message.name);
+    }
+
+    if (message.name === "VpnRecommender::action") {
+      eventName = "action";
+      this.openVpnPage();
+    }
+
+    if (message.name === "VpnRecommender::dismiss") {
+      eventName = "dismiss";
+    }
+
+    if (message.name === "VpnRecommender::info") {
+      eventName = "info";
+    }
+
+    if (message.name === "VpnRecommender::timeout") {
+      eventName = "timeout";
+    }
+
+    this.telemetryCallback({
+      "message_type": "event",
+      "event": eventName,
+      "dont-show-checked": String(message.data && message.data.dontShowChecked),
+    });
+  }
+
+  tryShowNotification() {
+    if (Date.now() - Number(Preferences.get(LAST_NOTIFICATION_PREF)) < TWENTY_FOUR_HOURS) {
+      log("less than 24 hours has passed since the last notification was shown");
+      return;
+    }
+
+    if (Preferences.get(DONT_SHOW_PREF)) return;
+
+    const notificationCount = Preferences.get(NOTIFICATION_COUNT_PREF) || 0;
+    if (notificationCount === MAX_NOTIFICATION_COUNT) return;
+
+    const doorhanger = new this.Doorhanger(this.notificationActionCallback.bind(this), `${this.extensionUrl}privileged/vpnRecommender`);
+    const success = doorhanger.present({message: DOORHANGER_MESSAGES[this.variation]});
+
+    if (!success) return;
+
+    Preferences.set(NOTIFICATION_COUNT_PREF, notificationCount + 1); // increment notification count
+    Preferences.set(LAST_NOTIFICATION_PREF, String(Date.now()));
+  }
+
+  addCleanUpFunction(func) {
+    this.cleanUpFunctions.push(func);
+  }
+
+  killNotification() {
```

Comment:
I don't understand why you're implementing this function twice. Couldn't you just call `Doorhanger.killNotification`? (Should probably also be called  `killAllNotifications`)

Reply:
I completely changed the way killing notifications work. Now there is a destruct() function that is called by doorhanger object itself after it concludes in one of three ways "auto-dismiss", "dismiss", and "action". In the case of study ending, api.js calls the destruct() function of the open doorhanger on clean up.